evol/problems/magicsanta.py

```diff
@@ -0,0 +1,67 @@
+import math
+from typing import List, Union
+
+from .problem import Problem
+from evol.helpers.utils import flatten, sliding_window
+
+
+class MagicSanta(Problem):
+    def __init__(self, city_coordinates, home_coordinate, gift_weight=None, sleigh_weight=1):
+        """
+        This problem is based on this kaggle competition: https://www.kaggle.com/c/santas-stolen-sleigh#evaluation.
+        :param distance_matrix:
+        :param gift_weight:
+        """
+        self.coordinates = city_coordinates
+        self.home_coordinate = home_coordinate
+        self.gift_weight = gift_weight
+        if gift_weight is None:
+            self.gift_weight = [1 for _ in city_coordinates]
+        self.sleigh_weight = sleigh_weight
+
+    @staticmethod
+    def distance(coord_a, coord_b):
+        return math.sqrt(sum([(z[0] - z[1]) ** 2 for z in zip(coord_a, coord_b)]))
+
+    def check_solution(self, solution: List[List[int]]):
+        """
+        Check if the solution for the problem is valid.
+        :param solution: List of lists containing integers representing visited cities.
+        :return: None, unless errors are raised.
+        """
+        set_visited = set(flatten(solution))
+        set_problem = set(range(len(self.coordinates)))
+        if set_visited != set_problem:
+            missing = set_problem.difference(set_visited)
+            extra = set_visited.difference(set_problem)
+            raise RuntimeError(f"Not all cities are visited! Missing: {missing} Extra: {extra}")
+        visited_cities = []
+        double_cities = []
+        for route in solution:
```

Comment:
Isn't it much more efficient here to use a `Counter` object?

Reply:
I wanted to be explicit in the error message. Both because I can test more explicitly (missing vs. extra) as well as inform the user a bit better on what is happening. 

When I used this problem during a training last year I noticed the students were having all sorts of errors and they were very happy with the explicit error msg.

Reply:
I don't think my explanation was clear enough, hence I committed my proposed solution. Is that acceptable for you? It should have the same result.